File: 2025/11/main.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"
)

func main() {
	tStart := time.Now()

	graph := GetInput("in.txt")
	sol1 := CountPaths(graph, "you", "out")
	sol2 := CountPathsInludingItems(graph, "svr", "out", "dac", "fft")

	fmt.Println(time.Since(tStart))

	fmt.Println(sol1)
	fmt.Println(sol2)
}

// Consider paths: (start -> a -> b -> end) OR (start -> b -> a -> end)
func CountPathsInludingItems(graph Graph, start, end, a, b string) int {
	paths1 := CountPaths(graph, start, a)
	paths1 *= CountPaths(graph, a, b)
	paths1 *= CountPaths(graph, b, end)

	paths2 := CountPaths(graph, start, b)
	paths2 *= CountPaths(graph, b, a)
	paths2 *= CountPaths(graph, a, end)

	return paths1 + paths2
}

func CountPaths(graph Graph, a, b string) int {
	sortedNodes := TopologicalSort(graph)

	paths := map[string]int{}
	for node := range graph {
		paths[node] = 0
	}

	paths[a] = 1

	for _, sortedNode := range sortedNodes {
		for _, node := range graph[sortedNode] {
			paths[node] += paths[sortedNode]
		}
	}

	return paths[b]
}

func TopologicalSort(graph Graph) []string {
	connections := map[string]int{}
	for node := range graph {
		connections[node] = 0
	}

	for _, nextNodes := range graph {
		for _, node := range nextNodes {
			connections[node]++
		}
	}

	zeroNodes := make([]string, 0)
	for node, count := range connections {
		if count == 0 {
			zeroNodes = append(zeroNodes, node)
		}
	}

	sortedNodes := make([]string, 0, len(graph))

	for len(zeroNodes) > 0 {
		currentNode := zeroNodes[0]
		zeroNodes = zeroNodes[1:]
		sortedNodes = append(sortedNodes, currentNode)

		for _, neighbor := range graph[currentNode] {
			connections[neighbor]--
			if connections[neighbor] == 0 {
				zeroNodes = append(zeroNodes, neighbor)
			}
		}
	}

	return sortedNodes
}

func NumberOfPaths(graph Graph, start, end string) int {
	paths := 0
	itemsToProcess := []string{start}

	for len(itemsToProcess) != 0 {
		item := itemsToProcess[0]
		itemsToProcess = itemsToProcess[1:]
		if item == end {
			paths++
			continue
		}

		nextItems := graph[item]
		itemsToProcess = append(itemsToProcess, nextItems...)
	}

	return paths
}

func GetInput(filename string) Graph {
	file, _ := os.ReadFile(filename)
	lines := strings.Split(string(file), "\n")

	graph := Graph{}
	for _, line := range lines {
		items := strings.Split(line, " ")
		key := items[0][:len(items[0])-1]
		graph[key] = items[1:]
	}

	return graph
}

func ShowGraph(graph Graph) {
	for k, v := range graph {
		fmt.Println(k, v)
	}
}

type GraphNode struct {
	item string
	prev *GraphNode
}

type Graph map[string][]string

File: 2025/11/main.py
from collections import defaultdict, deque
from time import time


def main():
    t1 = time()

    graph = get_input("in.txt")
    sol1 = count_paths(graph, "you", "out")
    sol2 = count_paths_including_items(graph, "svr", "out", "dac", "fft")

    t2 = time()

    print(f"Time: {((t2 - t1) * 1000):.6f} seconds")

    print(sol1)
    print(sol2)


def topological_sort(graph: dict[str, list[str]]) -> list[str]:
    in_degree = defaultdict(int)
    for u in graph:
        for v in graph[u]:
            in_degree[v] += 1

    zero_in_degree = deque([u for u in graph if in_degree[u] == 0])
    sorted_list = []

    while zero_in_degree:
        u = zero_in_degree.popleft()
        sorted_list.append(u)
        for v in graph[u]:
            in_degree[v] -= 1
            if in_degree[v] == 0:
                zero_in_degree.append(v)

    if len(sorted_list) != len(graph):
        raise ValueError("Graph has at least one cycle")

    return sorted_list


def count_paths(graph: dict[str, list[str]], a: str, b: str) -> int:
    sorted_nodes = topological_sort(graph)

    paths = {node: 0 for node in graph}
    paths[a] = 1

    for sorted_node in sorted_nodes:
        for node in graph[sorted_node]:
            paths[node] += paths[sorted_node]

    return paths[b]


def count_paths_including_items(
    graph: dict[str, list[str]], start: str, end: str, a: str, b: str
) -> int:
    paths1 = count_paths(graph, start, a)
    paths1 *= count_paths(graph, a, b)
    paths1 *= count_paths(graph, b, end)

    paths2 = count_paths(graph, start, b)
    paths2 *= count_paths(graph, b, a)
    paths2 *= count_paths(graph, a, end)

    return paths1 + paths2


def get_input(filename: str) -> dict[str, list[str]]:
    graph = defaultdict(list)
    with open(filename) as f:
        for line in f:
            items = line.strip().split(" ")
            graph[items[0][:-1]] = items[1:]
    return graph


if __name__ == "__main__":
    main()
